teuthology/provision/cloud/openstack.py

```diff
@@ -0,0 +1,331 @@
+            create_args['ex_security_groups'] = security_groups
+        self._node = self.provider.driver.create_node(
+            **create_args
+        )
+        log.debug("Created node: %s", self.node)
+        results = self.provider.driver.wait_until_running(
+            nodes=[self.node],
+        )
+        self._node, self.ips = results[0]
+        log.debug("Node started: %s", self.node)
+        self._create_volumes()
+        self._update_dns()
+        # Give cloud-init a few seconds to bring up the network, start sshd,
+        # and install the public key
+        time.sleep(20)
+        self._wait_for_ready()
+        return self.node
+
+    def _create_volumes(self):
+        vol_count = self.conf['volumes']['count']
+        vol_size = self.conf['volumes']['size']
+        name_templ = "%s_%0{0}d".format(len(str(vol_count - 1)))
+        vol_names = [name_templ % (self.name, i)
+                     for i in range(vol_count)]
+        try:
+            for name in vol_names:
+                volume = self.provider.driver.create_volume(
+                    vol_size,
+                    name,
+                )
+                log.info("Created volume %s", volume)
+                self.provider.driver.attach_volume(
+                    self.node,
+                    volume,
+                    device=None,
+                )
+        except Exception:
+            log.exception("Failed to create or attach volume!")
+            self._destroy_volumes()
+            return False
+        return True
+
+    def _destroy_volumes(self):
+        all_volumes = self.provider.driver.list_volumes()
+        our_volumes = [vol for vol in all_volumes
+                       if vol.name.startswith("%s_" % self.name)]
+        for vol in our_volumes:
+            try:
+                self.provider.driver.detach_volume(vol)
+            except Exception:
+                log.exception("Could not detach volume %s", vol)
+            try:
+                self.provider.driver.destroy_volume(vol)
+            except Exception:
+                log.exception("Could not destroy volume %s", vol)
+
+    def _update_dns(self):
+        query = urllib.urlencode(dict(
+            name=self.name,
+            ip=self.ips[0],
+        ))
+        nsupdate_url = "%s?%s" % (
+            config.nsupdate_url,
+            query,
+        )
+        resp = requests.get(nsupdate_url)
+        resp.raise_for_status()
+
+    def _wait_for_ready(self):
+        with safe_while(sleep=6, tries=20) as proceed:
+            while proceed():
+                try:
+                    self.remote.connect()
+                    break
+                except (
+                    socket.error,
+                    NoValidConnectionsError,
+                    AuthenticationException,
+                ):
+                    pass
+        cmd = "while [ ! -e '%s' ]; do sleep 5; done" % self._sentinel_path
+        self.remote.run(args=cmd, timeout=600)
```

Comment:
The way I'm reading this, this is paramiko's timeout, and it's a timeout for read or write operations.  I'm not sure this actually times out the command the way you want it to.  paramiko.client.exec_command(), timeout refers to Channel.settimeout.settimeout http://docs.paramiko.org/en/2.1/api/channel.html#paramiko.channel.Channel.settimeout.  Maybe the doc is misleading?